Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,500 @@
+            exponent = exponentSign * (Int(data[range]) ?? 0)
+        }
+        if let pointIndex = decimalPart.firstIndex(of: ".") {
+            // move the exponent by the number of digits after the decimal point
+            // so we are looking at an "integer" significand, easier to reason about
+            exponent -= decimalPart.distance(from: pointIndex, to: decimalPart.endIndex) - 1
+        }
+
+        while exponent > Self.exponentMax && digits.count <= Self.maxSignificandDigits {
+            // Clamping upper bound: Exponent is too large, try shifting zeros into the coefficient
+
+            digits.append(0)
+            exponent -= 1
+        }
+
+        while exponent < Self.exponentMin && !digits.isEmpty {
+            // Clamping lower bound: Exponent is too small, try taking zeros off the coefficient
+
+            if digits.count == 1 && digits[0] == 0 {
+                exponent = Self.exponentMin
+                break
+            }
+
+            if digits.last == 0 {
+                digits.removeLast()
+                exponent += 1
+                continue
+            }
+
+            if digits.last != 0 {
+                // We don't end in a zero and our exponent is too small
+                throw BSONError.InvalidArgumentError(message: "Underflow Error: \(data)")
+            }
+        }
+
+        guard exponent >= Self.exponentMin else {
+            throw BSONError.InvalidArgumentError(message: "Underflow Error: \(data)")
+        }
+
+        guard exponent <= Self.exponentMax else {
+            throw BSONError.InvalidArgumentError(message: "Overflow Error: \(data)")
+        }
+
+        guard digits.count <= Self.maxSignificandDigits else {
+            throw BSONError.InvalidArgumentError(message: "Overflow Error: \(data)")
+        }
+
+        let significandLoDigits = [UInt8](digits.suffix(Self.maxSignificandDigits / 2)).decimalDigitsToUInt64()
+        let significandHiDigits = [UInt8](digits.dropLast(Self.maxSignificandDigits / 2)).decimalDigitsToUInt64()
+
+        // Multiply by one hundred quadrillion (note the seventeen zeroes)
+        // the product is the significandHiDigits "shifted" up by 17 decimal places
+        // we can then add the significandLoDigits to the product to ensure that we have a correctly formed significand
+        let product = significandHiDigits.multipliedFullWidth(by: Self.decimalShift17Zeroes)
+        var significand = UInt128(hi: product.high, lo: product.low)
+
+        let (result, didOverflow) = significand.lo.addingReportingOverflow(significandLoDigits)
+        significand.lo = result
+
+        if didOverflow {
+            significand.hi += 1
+        }
+
+        let biasedExponent = UInt64(exponent + Self.exponentBias).getLeastSignificantBits(Self.exponentLength)
+
+        var value = UInt128()
+
+        // Normally Decimal(K) encodings would conditionally modify the combination field here
```

Comment:
one caveat to this is that we're supposed to treat significands that are too big as 0 according to the IEEE spec. According to the MongoDB spec, we're supposed to error in overflow/underflow scenarios though. So are we covered here?

Reply:
I believe so, if I understand correctly it isn't possible to formulate a number in string form that in decimal128 encoding will encounter that case of the leading 0b11